src/engine/AbsoluteValueConstraint.cpp

```diff
@@ -0,0 +1,454 @@
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#include "AbsoluteValueConstraint.h"
+#include "ConstraintBoundTightener.h"
+#include "Debug.h"
+#include "FloatUtils.h"
+#include "ITableau.h"
+#include "MStringf.h"
+#include "MarabouError.h"
+#include "PiecewiseLinearCaseSplit.h"
+#include "Statistics.h"
+
+AbsoluteValueConstraint::AbsoluteValueConstraint( unsigned b, unsigned f )
+    : _b( b )
+    , _f( f )
+    , _haveEliminatedVariables( false )
+{
+    setPhaseStatus( PhaseStatus::PHASE_NOT_FIXED );
+}
+
+PiecewiseLinearConstraint *AbsoluteValueConstraint::duplicateConstraint() const
+{
+    AbsoluteValueConstraint *clone = new AbsoluteValueConstraint( _b, _f );
+    *clone = *this;
+    return clone;
+}
+
+void AbsoluteValueConstraint::restoreState( const PiecewiseLinearConstraint *state )
+{
+    const AbsoluteValueConstraint *abs = dynamic_cast<const AbsoluteValueConstraint *>( state );
+    *this = *abs;
+}
+
+void AbsoluteValueConstraint::registerAsWatcher( ITableau *tableau )
+{
+    tableau->registerToWatchVariable( this, _b );
+    tableau->registerToWatchVariable( this, _f );
+}
+
+void AbsoluteValueConstraint::unregisterAsWatcher( ITableau *tableau )
+{
+    tableau->unregisterToWatchVariable( this, _b );
+    tableau->unregisterToWatchVariable( this, _f );
+}
+
+void AbsoluteValueConstraint::notifyVariableValue( unsigned variable, double value )
+{
+    _assignment[variable] = value;
+}
+
+/*
+  The variable watcher notification callbacks, about a change in a variable's value or bounds.
+  suppose A < x_b < B, C < x_f < D
+  if variable == x_b then:
+    A < 0 & B < 0 then: max{|B|, C} < x_f < min{|A|, D}
+    A < 0 & B > 0 then: max{0, C} < x_f < min{max{|A|, B}, D}
+    A > 0 & B < 0 then: ------
+    A > 0 & B > 0 then: max{A, C} < x_f < min{B, D}
+  if variable == x_f then:
+    C > 0 & D > 0 then: min{max{-D, A}, max{A, C}} < x_b < max{min{-C, A}, min{B, D}}
+*/
+
+void AbsoluteValueConstraint::notifyLowerBound( unsigned variable, double bound )
+{
+    if ( _statistics )
+        _statistics->incNumBoundNotificationsPlConstraints();
+
+    if ( _lowerBounds.exists( variable ) &&
+         !FloatUtils::gt( bound, _lowerBounds[variable] ) )
+        return;
+
+    _lowerBounds[variable] = bound;
+
+    // Check whether the phase has become fixed
+    fixPhaseIfNeeded();
+
+    // Update partner's bound
+    if ( isActive() && _constraintBoundTightener )
+    {
+        if ( variable == _b )
+        {
+            if ( bound < 0 )
+            {
+                double fUpperBound = FloatUtils::max( -bound, _upperBounds[_b] );
+                _constraintBoundTightener->registerTighterUpperBound( _f, fUpperBound );
+            }
+            else
+            {
+                // Phase is fixed, don't care about this case
```

Comment:
Is this always going to be the case? From the perspective of generating explanations in the future.

Reply:
Yes, this is here only for understandability. If the phase has become fixed (elsewhere), that the explanations should be generated elsewhere, also.